### PR TITLE
Reduce the amount of bad Docker.inspects when using docker_only flag in combination with Systemd

### PR DESCRIPTION
**Pull request by basvdlei**
When running with the `docker_only` flag, the manager will keep trying to find a handler for the "ignored" raw containers during the housekeeping (every minute.) Each time a container is passed to the docker driver it will send an inspect request to the Docker daemon, resulting in loads of log messages:

```log
Jul 18 19:18:45 hactar docker[1024]: time="2015-07-18T19:18:45+02:00" level=info msg="+job container_inspect(rsyslog.service)"
Jul 18 19:18:45 hactar docker[1024]: no such id: rsyslog.service
Jul 18 19:18:45 hactar docker[1024]: time="2015-07-18T19:18:45+02:00" level=info msg="-job container_inspect(rsyslog.service) = ERR (1)"
Jul 18 19:18:45 hactar docker[1024]: time="2015-07-18T19:18:45+02:00" level=error msg="Handler for GET /containers/{name:.*}/json returned error: no such id: rsyslog.service"
Jul 18 19:18:45 hactar docker[1024]: time="2015-07-18T19:18:45+02:00" level=error msg="HTTP Error: statusCode=404 no such id: rsyslog.service"
```

This PR adds an extra check in the CanHandleAndAccept method of the Docker driver to filter out non-Docker containers when using Systemd before doing a Docker inspect.

### CODE REVIEW

File: container/docker/factory.go [hunk 1]
```diff
@@ -139,6 +139,12 @@ func FullContainerName(dockerId string) string {
 func (self *dockerFactory) CanHandleAndAccept(name string) (bool, bool, error) {
 	// docker factory accepts all containers it can handle.
 	canAccept := true
+
+	// When using SystemD all docker containers have a docker- prefix
+	if UseSystemd() && !strings.HasPrefix(path.Base(name), "docker-") {
+		return false, canAccept, nil
```
Comment (rjnagal) [inline container/docker/factory.go]: I think this might not be true if a docker container is started with --cgroup_parent flag. parent cgroup name will replace docker in that case.

Comment (basvdlei) [inline container/docker/factory.go]: You are correct. I've not used that feature before :) So I just did a simple test and the prefix `docker-` is indeed replaced with the parent name. But it also didn't show up as a docker container in a build of master though. Probably because the docker factory is already relying on the `docker-` prefix when Systemd is detected :(

**Comment by basvdlei** [comment]
I'm pretty sure containers are started as a `.scope` unit. This is good enough for filtering on my systemd based machines (Debian and CoreOS), where I don't have any other scope units. I've also changed the ContainerNameToDockerId to handle containers started with `--cgroup-parent`.

**Comment by basvdlei** [comment]
Hi, is there something I can do to get this in?

**Comment by whosthatknocking** [comment]
Just wondering, it looks like the scope of this PR is,
  - enabling support for `--cgroup-parent`
  - filter accept to only `.scope`

Perhaps these should be,
  - Two PRs?
  - To minimize any performance impact should the `regexp.MustCompile` be moved outside the func

**Comment by vishh** [comment]
ok to test

**Comment by vishh** [comment]
LGTM

File: container/docker/factory.go [hunk 2]
```diff
@@ -118,8 +118,12 @@ func ContainerNameToDockerId(name string) string {
 
 	// Turn systemd cgroup name into Docker ID.
 	if UseSystemd() {
-		id = strings.TrimPrefix(id, "docker-")
-		id = strings.TrimSuffix(id, ".scope")
+		// Using a regexp because container started with
+		// --cgroup-parent have another prefix than 'docker'
+		re := regexp.MustCompile(`[A-z]+-([a-z0-9]+)\.scope`)
```
Comment (jimmidyson) [inline container/docker/factory.go]: As @whosthatknocking said, please move this to a var outside of the func.

**Comment by basvdlei** [comment]
Moved the regexp out of the function as requested. Squashed and rebased so the automated builds won't fail (on changed dependencies.)

**Comment by jimmidyson** [comment]
How should the non-docker cgroup parent be handled in `FullContainerName`?

**Comment by basvdlei** [comment]
The first thing that comes to mind is to ask Docker for the cgroup parent name. It seems to be part of the inspect result. The current `FullContainerName` implementation however doesn't have a reference to the Docker client.

When scanning the codebase, `FullContainerName` is only used in `dockerContainerHandler.ListContainers`. So adding an extra argument that passes a Docker client reference shouldn't have that big of an impact.

But I don't like the amount of extra API call to Docker this will generate, for a case most users won't ever need. 

We could first try to test if `docker-%s.scope` cgroup exists before bugging Docker about to cgroup parent name?

File: container/docker/factory.go [hunk 3]
```diff
@@ -42,6 +42,10 @@ var DockerNamespace = "docker"
 var dockerRootDir = flag.String("docker_root", "/var/lib/docker", "Absolute path to the Docker state root directory (default: /var/lib/docker)")
 var dockerRunDir = flag.String("docker_run", "/var/run/docker", "Absolute path to the Docker run directory (default: /var/run/docker)")
 
+// Regexp that identifies docker cgroups, containers started with
+// --cgroup-parent have another prefix than 'docker'
+var dockerCgroupRegexp = regexp.MustCompile(`[A-z]+-([a-z0-9]+)\.scope`)
```
Comment (jimmidyson) [inline container/docker/factory.go]: This regex can probably be locked to `([a-z0-9]{64})\.scope$`, including anchoring to end. I wonder if we could use that to get the parent cgroup too? Something like `(.*)-([a-z0-9]{64})\.scope$`?

Comment (whosthatknocking) [inline container/docker/factory.go]: `(.+)-([a-z0-9]{64})\.scope$`?

File: container/docker/factory.go [hunk 4]
```diff
@@ -140,6 +145,12 @@ func FullContainerName(dockerId string) string {
 func (self *dockerFactory) CanHandleAndAccept(name string) (bool, bool, error) {
 	// docker factory accepts all containers it can handle.
 	canAccept := true
+
+	// When using Systemd all docker containers have a .scope suffix
+	if UseSystemd() && !strings.HasSuffix(path.Base(name), ".scope") {
```
Comment (whosthatknocking) [inline container/docker/factory.go]: Feels appropriate to follow existing format and express the check under `isContainerName` or similar function?

File: container/docker/factory.go (hunk 3 above)
Comment (jimmidyson) [inline container/docker/factory.go]: AFAIK format is `<cgroupParent>-<containerId>.scope`. Container ID is 64 characters (hence the `{64}` so from this regex we should be able to extract the cgroup parent (normally `docker`) & the container ID. I can't imagine any conflict with other scope units so this should be pretty safe.

**Comment by jimmidyson** [comment]
@basvdlei Are you still working on this PR? If not do you mind if I pick this up & swap over the regex? Would really like to get this in - nice optimisation.

**Comment by basvdlei** [comment]
Haven't spend much time on it, and when I did, I've mostly been banging my head on finding a clean solution for `FullContainerName`. The more I look at it, the more it seems like dead code. I'll take a look at your PR in a moment.

I like whosthatknocking's regex `(.+)` for the parent name, since it's never nil. AFAIK docker does not place any restrictions on the name, as far as it results in a valid unit name.

File: container/docker/factory.go (hunk 3 above)
Comment (jimmidyson) [inline container/docker/factory.go]: Ah I missed @whosthatknocking's `+` as opposed to my `*` - much better idea. Will update #896 with that.

File: container/docker/handler.go [hunk 5]
```diff
@@ -294,32 +294,8 @@ func (self *dockerContainerHandler) GetStats() (*info.ContainerStats, error) {
 }
 
 func (self *dockerContainerHandler) ListContainers(listType container.ListType) ([]info.ContainerReference, error) {
-	if self.name != "/docker" {
```
Comment (jimmidyson) [inline container/docker/handler.go]: How does removing this affect non-systemd Docker installs?

Comment (basvdlei) [inline container/docker/handler.go]: In all the tests I've done, it doesn't. I have tested this on Ubuntu Precise (upstart), Debian Jessie/Sid (sysv init) and Debian Jessie (systemd).

When the manager is started it registers the `/` container and starts searching for containers using `detectSubcontainers("/")`. Since `/` is technically a raw container, all Docker containers are discovered using their entries in the cgroup mount. 

I have not found any place where the `dockerContainerHandler` with the name `/docker` is created. Because it doesn't exist, this **if** will never be **false**

Comment (jimmidyson) [inline container/docker/handler.go]: Coincentally I've just run cadvisor on Ubnutu 14.04 & it does actually use `/docker`...

```
condescending_bhabha (/docker/f5a39e9c98788da89c050ca45d266ce0ddc31bd0a7e3ecbc618e166c17ade17d)
```

Comment (basvdlei) [inline container/docker/handler.go]: Yes, docker will create a cgroup named `docker`, and is found as subcontainer of `/`. But `CanHandleAndAccept` will be **false** when the manager tries to get a `dockerContainerHandler` for it. Because the inspect will fail: https://github.com/google/cadvisor/blob/master/container/docker/factory.go#L147

You can also see this happening in the Docker daemon logs:
```log
ERRO[437132] Handler for GET /containers/{name:.*}/json returned error: no such id: docker 
ERRO[437132] HTTP Error                                    err=no such id: docker statusCode=404
```
Only when the container `/docker/f5a39e9c98788da89c050ca45d266ce0ddc31bd0a7e3ecbc618e166c17ade17d` is found it will actually create a handler.

Comment (jimmidyson) [inline container/docker/handler.go]: The docker ID is extracted from the container name - https://github.com/google/cadvisor/blob/499737d7ad3f754a63036e9c4d0d5f77f2889fed/container/docker/factory.go#L118. It is that ID that is used in the docker inspect.

Again, this is tested & working properly on Ubuntu 14.04.

Comment (basvdlei) [inline container/docker/handler.go]: Maybe we are saying the same thing. Or are you saying this code is still needed? If so, do you know where the `dockerContainerHandler` named *only* `/docker` is created? Because only then will `ListContainers` actually do something.

In my tests, with this function stubbed, everything also still works on a non-systemd machine and also see Docker containers with the `/docker` cgroup eg.

```
clever_wright (/docker/64f26ce9d60219c5d730f31abeabc39c47fe7aa282343c3a3a8c820a80db5b70)
```

Comment (jimmidyson) [inline container/docker/handler.go]: omg you're totally right - sorry i misread the condition somehow. Doh!

Yes I agree - redundant code to be removed.

File: container/docker/factory.go [hunk 6]
```diff
@@ -119,27 +123,24 @@ func ContainerNameToDockerId(name string) string {
...
+		if matches := dockerCgroupRegexp.FindStringSubmatch(id); matches != nil {
+			id = matches[1]
+		}
 	}
 
 	return id
 }
 
-// Returns a full container name for the specified Docker ID.
-func FullContainerName(dockerId string) string {
```
Comment (jimmidyson) [inline container/docker/factory.go]: Why is this deleted? Isn't this used by the web pages?

Comment (jimmidyson) [inline container/docker/factory.go]: OK it's redundant too. I like this PR - lots deleted :)

**Comment by jimmidyson** [comment]
@vishh LGTM but need to retest - not sure if i'm in the admin list in Jenkins to request that with a comment here?

File: container/docker/factory.go (hunk 4 above)
Comment (jimmidyson) [inline container/docker/factory.go]: Agreed. Also using the regexp rather than `HasSuffix` to check the container name would make more sense.

**Comment by jimmidyson** [comment]
LGTM - thanks!
